### PR TITLE
Remix of PR #420

### PR DESCRIPTION
**Pull request by fevangelista**
This is a reworked version of PR #420. Here is the original description:

## Description
This PR implements new orbital orthonormalization procedure for initial orbitals in Forte.
The new procedure keeps the relative ordering of the guess orbitals, that is, no canonicalization procedure is performed.
Now initial orbitals can come from a different basis set (see added test case `df-casscf-4-basis`).

To actually check these initial guess orbitals, some Python codes can be added in the input:
```
molecule Cr2{
Cr
Cr 1 3.5
symmetry c1
}

set globals{
scf_type             df
reference            rhf
guess                sadno
basis                cc-pwcvtz
df_basis_scf         def2-universal-jkfit
df_basis_mp2         def2-universal-jkfit
restricted_docc       [18]
active                [22]
}
escf, wfn = energy('scf', return_wfn=True)

from forte._forte import make_mo_space_info_from_map
mo_space_info = make_mo_space_info_from_map(wfn.nmopi(),
                                            Cr2.point_group().symbol(),
                                            {"ACTIVE": [22], "RESTRICTED_DOCC": [18]}, [])

# same basis, different geometry
wfn_old = wfn.from_file('wfn1.npy')  # cc-pwCVTZ @ 3.25 Angstrom
from forte.proc.orbital_helpers import ortho_orbs_new
wfn.Ca().copy(ortho_orbs_new(wfn.Ca(), wfn.S(), wfn_old.Ca(), mo_space_info))

# different basis, different geometry
wfn_old = wfn.from_file('wfn2.npy')  # cc-pVDZ @ 3.25 Angstrom
from forte.proc.orbital_helpers import basis_projection
Ca = basis_projection(wfn_old, wfn, mo_space_info)
wfn.Ca().copy(Ca)

# plot orbitals
orbs = [19 + i for i in range(22)]
set cubeprop_orbitals $orbs
cubeprop(wfn)
```

## User Notes
- [x] Add new orbital projection code that works with symmetry.
- [x] Can now use this code to restart computations from a different geometry and using a different basis.
- [x] No changes to compilation

## Checklist
- [x] Added/updated tests of new features and included a reference `output.ref` file
- [x] Removed comments in code and input files
- [x] Documented source code
- [x] Checked for redundant headers
- [x] Checked for consistency in the formatting of the output file
- [x] Documented new features in the manual
- [x] Ready to go!

### CODE REVIEW

File: forte/modules/objects_factory_psi4.py [hunk 1]
```diff
@@ -18,8 +18,7 @@
 
 from forte.data import ForteData
 
-from forte.proc.orbital_helpers import orbital_projection
-from forte.proc.orbital_helpers import read_orbitals, ortho_orbs_forte
+from forte.proc.orbital_helpers import read_orbitals, basis_projection, orbital_projection, ortho_orbs_new
```
Comment (lcyyork) [inline forte/modules/objects_factory_psi4.py:21]:
```suggestion
```

File: forte/modules/objects_factory_psi4.py [hunk 2]
```diff
@@ -97,29 +90,31 @@ def prepare_forte_objects(self, data, **kwargs):
...
-        kwargs.pop("molecule", None)
+        elif "basis" in kwargs and basis != "":
+            raise ValueError("Both basis set in options and kwargs are provided. Please provide only one.")
 
-        # Step 2. Get or compute the reference wavefunction
-        psi_wfn, is_psi_wfn_fresh = self.get_psi4_wavefunction(data, **kwargs)
-        kwargs.pop("ref_wfn", None)
-        data.psi_wfn = psi_wfn
+        if "basis" not in kwargs:
+            kwargs["basis"] = basis
```
Comment (lcyyork) [inline forte/modules/objects_factory_psi4.py:108]:
```
if "basis" not in kwargs:
    basis = data.options.get_str("BASIS")
    if not basis:
        basis = psi4.core.get_global_option("BASIS")
        p4print(f" Using basis set {basis} from Psi4 global options.")
    kwargs["basis"] = basis
else:
    if data.options.get_str("BASIS"):
        raise ValueError("Both basis set in options and kwargs are provided. Please provide only one.")
```

File: forte/modules/objects_factory_psi4.py [hunk 3]
```diff
@@ -135,55 +130,85 @@ def prepare_forte_objects(self, data, **kwargs):
...
+                if norm < 1e-10:
+                    return ref_wfn
+
+            # compute a new Wavefunction object
+            new_wfn = self.run_psi4("scf", data.molecule, False, **kwargs)
+
+            pCa = new_wfn.basis_projection(ref_wfn.Ca(), ref_wfn.nmopi(), ref_wfn.basisset(), new_wfn.basisset())
+            pCb = new_wfn.basis_projection(ref_wfn.Cb(), ref_wfn.nmopi(), ref_wfn.basisset(), new_wfn.basisset())
+            new_wfn.guess_Ca(pCa)
+            new_wfn.guess_Cb(pCb)
```
Comment (lcyyork) [inline forte/modules/objects_factory_psi4.py:207]: These are passed to `guess_Ca_` and `guess_Cb_` in SCF, but I thought we read orbitals from `Ca_`.

**Comment by lcyyork** [comment]
Thank you for clearing up the logic. I only have a few minor comments.

File: forte/modules/objects_factory_psi4.py (hunk 3 above)
Comment (fevangelista) [reply forte/modules/objects_factory_psi4.py]: Thanks for catching this!

**Comment by lcyyork** [comment]
@fevangelista This new procedure fails if the target basis set is smaller than the one in the reference. For example, `energy('forte', ref_wfn=wfn, basis='cc-pvdz')` fails if `wfn` is obtained from cc-pVTZ. The dimension of `Ca` is incorrect.

I also have a concern of doing general basis set projection without separating occupied and unoccupied. Is the ordering of occupied orbitals preserved between the two sets of orbitals? This seems important for maintaining a good guess of active space for CASSCF.

**Pull request merged**